File: naive_route.py


from typing import List, Dict, Set
from collections import defaultdict
from ordered_set import OrderedSet  # optional if order matters
import heapq

from dijkstra import DijkstraModel
from grid import Grid, Net, Point, PointPair
from prim_mst import prim_mst

# Assuming these classes are defined elsewhere in your Python code
# Point, Grid, Net, PointPair, DijkstraModel, DijkstraResult

def naive_route(unrouted_grid: Grid) -> Grid:
    unrouted_grid.traces.clear()
    unrouted_grid.diagonal_traces.clear()
    
    def prepare_dijkstra_model_unrouted(net: Net, start: Point, end: Point) -> DijkstraModel:
        other_pads = unrouted_grid.pads_except(net)
        return DijkstraModel(
            width=unrouted_grid.width,
            height=unrouted_grid.height,
            obstacles=other_pads,
            diagonal_obstacles=set(),  # no diagonal obstacles in the unrouted grid
            start=start,
            end=end
        )
    
    # Prepare all the pairs of pads to route
    pad_pairs = []
    for net, points in unrouted_grid.pads.items():
        points_list = list(points)
        # Permutate all pairs and calculate their distance using Dijkstra's algorithm
        pairs = []
        for i in range(len(points_list)):
            for j in range(i + 1, len(points_list)):
                point1 = points_list[i]
                point2 = points_list[j]
                dijkstra_model = prepare_dijkstra_model_unrouted(net, point1, point2)
                try:
                    dijkstra_result = dijkstra_model.run()
                    distance = dijkstra_result.distance
                except:
                    distance = float('inf')
                pairs.append((distance, net, PointPair(point1, point2)))
        
        # Prim's algorithm (assuming prim_mst is implemented elsewhere)
        pairs = prim_mst(pairs)
        pad_pairs.extend(pairs)
    
    # Create a priority queue (min-heap)
    priority_queue = []
    items_by_distance = defaultdict(list)
    
    for distance, net, point_pair in pad_pairs:
        heapq.heappush(priority_queue, distance)
        items_by_distance[distance].append((net, point_pair))
    
    def prepare_dijkstra_model(net: Net, start: Point, end: Point) -> DijkstraModel:
        other_pads = unrouted_grid.pads_except(net)
        other_routes = unrouted_grid.routes_except(net)
        other_diagonal_routes = unrouted_grid.diagonal_routes_except(net)
        obstacles = other_pads.union(other_routes)
        return DijkstraModel(
            width=unrouted_grid.width,
            height=unrouted_grid.height,
            obstacles=obstacles,
            diagonal_obstacles=other_diagonal_routes,
            start=start,
            end=end
        )
    
    while priority_queue:
        distance = heapq.heappop(priority_queue)
        if not items_by_distance[distance]:
            continue  # skip if we've already processed all items with this distance
        net, point_pair = items_by_distance[distance].pop(0)
        
        # Construct dijkstra model for the current pair of pads
        dijkstra_model = prepare_dijkstra_model(net, point_pair.start, point_pair.end)
        
        # Run dijkstra's algorithm
        try:
            dijkstra_result = dijkstra_model.run()
            directions = dijkstra_result.directions
        except Exception as e:
            raise RuntimeError(f"Failed to run Dijkstra's algorithm: {e}")
        
        # Add the route to the grid
        current_point = point_pair.start
        if net not in unrouted_grid.traces:
            unrouted_grid.traces[net] = set()
        unrouted_grid.traces[net].add(current_point)
        
        for direction in directions:
            last_point = current_point
            current_point = Point(
                x=(current_point.x + direction.x),
                y=(current_point.y + direction.y)
            )
            assert current_point.x < unrouted_grid.width and current_point.y < unrouted_grid.height, "Point out of bounds"
            unrouted_grid.traces[net].add(current_point)
            
            if direction.x != 0 and direction.y != 0:
                # If the direction is diagonal, add the diagonal trace
                diagonal_trace_point = Point(
                    x=min(last_point.x, current_point.x),
                    y=min(last_point.y, current_point.y)
                )
                if net not in unrouted_grid.diagonal_traces:
                    unrouted_grid.diagonal_traces[net] = set()
                unrouted_grid.diagonal_traces[net].add(diagonal_trace_point)
    
    assert not priority_queue or all(not items for items in items_by_distance.values())
    return unrouted_grid

File: gui_leptos_tauri/src-tauri/src/naive_route.rs
use std::cmp::Reverse;
use std::collections::BTreeSet;
use std::collections::BinaryHeap;
use std::collections::HashSet;

use crate::dijkstra::*;
use crate::grid::*;
use crate::prim_mst::prim_mst;
use crate::proba_grid::TracePath;
use ordered_float::OrderedFloat;

pub fn naive_route(mut unrouted_grid: Grid) -> Result<Grid, String> {
    unrouted_grid.traces.clear();
    unrouted_grid.diagonal_traces.clear();
    let prepare_dijkstra_model_unrouted = |net: Net, start: Point, end: Point| {
        let other_pads = unrouted_grid.pads_except(&net);
        DijkstraModel {
            width: unrouted_grid.width,
            height: unrouted_grid.height,
            obstacles: other_pads,
            diagonal_obstacles: HashSet::new(), // no diagonal obstacles in the unrouted grid
            start,
            end,
        }
    };
    // prepare all the pairs of pads to route
    let pad_pairs: Vec<(OrderedFloat<f64>, Net, PointPair)> = unrouted_grid
        .pads
        .iter()
        .flat_map(|(net, points)| {
            // permutate all pairs and calculate their distance using Dijkstra's algorithm
            let mut pairs = vec![];
            let points_vec: Vec<Point> = points.iter().cloned().collect();
            for i in 0..points_vec.len() {
                for j in (i + 1)..points_vec.len() {
                    let point1 = points_vec[i];
                    let point2 = points_vec[j];
                    let dijkstra_model =
                        prepare_dijkstra_model_unrouted(net.clone(), point1, point2);
                    let DijkstraResult { distance, .. } =
                        dijkstra_model.run().unwrap_or(DijkstraResult {
                            start: point1,
                            end: point2,
                            trace_directions: vec![],
                            distance: f64::INFINITY,
                            trace_path: TracePath {
                                covered: BTreeSet::new(),
                                diagonal_covered: BTreeSet::new(),
                            },
                        });
                    pairs.push((
                        OrderedFloat(distance),
                        net.clone(),
                        PointPair::new(point1, point2),
                    ));
                }
            }

            // prim's algorithm
            let pairs = prim_mst(pairs);
            pairs
        })
        .collect();
    let mut priority_queue: BinaryHeap<_> = pad_pairs.into_iter().map(Reverse).collect();

    fn prepare_dijkstra_model(grid: &Grid, net: &Net, start: Point, end: Point) -> DijkstraModel {
        let other_pads = grid.pads_except(net);
        let other_routes = grid.routes_except(net);
        let other_diagonal_routes = grid.diagonal_routes_except(net);
        let mut obstacles = other_pads;
        obstacles.extend(other_routes);
        DijkstraModel {
            width: grid.width,
            height: grid.height,
            obstacles,
            diagonal_obstacles: other_diagonal_routes,
            start,
            end,
        }
    }
    let mut grid = unrouted_grid.clone();
    while let Some(Reverse((OrderedFloat(_distance), net, point_pair))) = priority_queue.pop() {
        println!(
            "Routing net: {:?}, from {:?} to {:?}",
            net,
            point_pair.start(),
            point_pair.end()
        );
        // construct dijkstra model for the current pair of pads
        let dijkstra_model =
            prepare_dijkstra_model(&grid, &net, point_pair.start(), point_pair.end());
        // run dijkstra's algorithm
        let DijkstraResult {
            trace_directions, ..
        } = dijkstra_model.run()?;
        // add the route to the grid
        let mut current_point = point_pair.start();
        grid.traces
            .entry(net.clone())
            .or_default()
            .insert(current_point);
        for direction in trace_directions {
            let last_point = current_point;
            current_point = Point {
                x: (current_point.x as i32 + direction.x) as usize,
                y: (current_point.y as i32 + direction.y) as usize,
            };
            assert!(
                current_point.x < grid.width && current_point.y < grid.height,
                "Point out of bounds"
            );
            grid.traces.get_mut(&net).unwrap().insert(current_point);
            if direction.x != 0 && direction.y != 0 {
                // if the direction is diagonal, we also add the diagonal trace
                let diagonal_trace_point = Point {
                    x: last_point.x.min(current_point.x),
                    y: last_point.y.min(current_point.y),
                };
                grid.diagonal_traces
                    .entry(net.clone())
                    .or_default()
                    .insert(diagonal_trace_point);
            }
        }
    }
    assert!(priority_queue.is_empty());
    Ok(grid)
}
